<a href="https://colab.research.google.com/github/Hojiakbardevs/Datascience/blob/main/Abdulhakimov_Hojiakbar_imtihon_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pandas kutubxonasiga doir bilimlaringizni mustaxkamlash uchun mashqlar!

Hurmatli o'quvchi, bilimingizni yanada mustaxkamlashingiz va chuqurroq ko'nikmalarga ega bo'lishingiz uchun sizga topshiriqlar tayyorladik. Sizga yengillik yaratish maqsadida topshiriqlarni `📌Mavzulashtirdik`

## `📌Pandas kutubxonasini yuklash`

**1.** Pandas kutubxonasini `pd` nomi ostida yuklang va pandas kutubxonasini chop eting.

In [ ]:
import pandas as pd

**2.** Versiya va konfiguratsiyani tekshiring.

In [ ]:
pd.__version__

'2.0.3'

**3.** Pandas kutubxonasida mavjud bo'lgan barcha versiya ma'lumotlarini chop eting.

In [ ]:
pd.show_versions

<function pandas.util._print_versions.show_versions(as_json: 'str | bool' = False) -> 'None'>

## `📌DataFrame: Asoslar`

DataFrames-da ma'lumotlarni tanlash(selecting), saralash(sorting), qo'shish(adding) va yig'ish(aggregating) uchun bir nechta asosiy mashqlar.

💡Eslatma: NumPy kutubxonasini import qilishni unutmang:
``` python
import numpy as np
```

Quyidagi Python lug'ati(dictionary) `data` va Python ro'yxati(list) `labels` ni ko'rib chiqing:


``` python
data = {'animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'],
        'age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
        'visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
        'priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']
```



In [ ]:
import numpy as np

In [ ]:
data = {'animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'],
        'age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
        'visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
        'priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

df = pd.DataFrame(data, index=labels)
df

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  2.0       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

**4.** Ushbu `lug'at`(dictionary) yordamida `data` dan `labels` indeksiga ega DataFrame yarating.

In [ ]:
import numpy as np

data = {'animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'],
        'age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
        'visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
        'priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

df = pd.DataFrame(data, index=labels)
df

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  2.0       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

**5.** Ushbu DataFrame va uning ma'lumotlari haqidagi asosiy ma'lumotlarning qisqacha mazmunini ko'rsating.

`💡Ishora:` DataFrame-da chaqirilishi mumkin bo'lgan yagona usul mavjud.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10 entries, a to j
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   animal    10 non-null     object 
 1   age       8 non-null      float64
 2   visits    10 non-null     int64  
 3   priority  10 non-null     object 
dtypes: float64(1), int64(1), object(2)
memory usage: 400.0+ bytes


In [ ]:
df.describe()

age     visits
count  8.000000  10.000000
mean   3.437500   1.900000
std    2.007797   0.875595
min    0.500000   1.000000
25%    2.375000   1.000000
50%    3.000000   2.000000
75%    4.625000   2.750000
max    7.000000   3.000000

**6.** DataFrame `df` dastlabki 3 qatorini qaytaring.

In [ ]:
df.head(3)

FlightNumber           Airline    From         To  delay_1  delay_2  \
0         10045               KLM  London      Paris     23.0     47.0   
1         10055     Air France 12  Madrid      Milan      NaN      NaN   
2         10065  British Airways   London  Stockholm     24.0     43.0   

   delay_3  
0      NaN  
1      NaN  
2     87.0

In [ ]:
df.iloc[:3]

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no

**7.** DataFrame `df`-dan faqat `animal` va `age` ustunlarini tanlang.

In [ ]:
df.loc[:, ['animal', 'age']]

animal  age
a    cat  2.5
b    cat  3.0
c  snake  0.5
d    dog  NaN
e    dog  5.0
f    cat  2.0
g  snake  4.5
h    cat  NaN
i    dog  7.0
j    dog  3.0

In [ ]:
df[['animal', 'age']]

animal  age
a    cat  2.5
b    cat  3.0
c  snake  0.5
d    dog  NaN
e    dog  5.0
f    cat  2.0
g  snake  4.5
h    cat  NaN
i    dog  7.0
j    dog  3.0

**8.** `[3, 4, 8]` va `['animal', 'age']` qatorlaridagi ma'lumotlarni tanlang.

In [ ]:
df.loc[df.index[[3, 4, 8]], ['animal', 'age']]

animal  age
d    dog  NaN
e    dog  5.0
i    dog  7.0

**9.** Faqat `visits` soni 3 tadan ko'p bo'lgan qatorlarni tanlang.

In [ ]:
df[df['visits']>2]

animal  age  visits priority
b    cat  3.0       3      yes
d    dog  NaN       3      yes
f    cat  2.0       3       no

10. Yosh yo'qolgan qatorlarni tanlang, ya'ni `NaN`

In [ ]:
df[df['age'].isnull()]

animal  age  visits priority
d    dog  NaN       3      yes
h    cat  NaN       1      yes

**11.** `animal` qatoridagi `cat` va `age` qatoridagi `3` dan kichik bo'lgan umumiy qatorlarni tanlang.

In [ ]:
df[(df['animal']=='cat') & (df['age']<3)]

animal  age  visits priority
a    cat  2.5       1      yes
f    cat  2.0       3       no

**12.** `age` 2 va 4 (shu jumladan) orasidagi qatorlarni tanlang.

In [ ]:
df[df['age'].between(2, 4)]

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
f    cat  2.0       3       no
j    dog  3.0       1       no

**13.** `f` qatoridagi yoshni `1,5` ga o'zgartiring.

In [ ]:
df.loc['f', 'age'] = 1.5

**14.** `df` yordamida `visits` qatorining yig'indisini hisoblang.

In [ ]:
df['visits'].sum()

19

**15.** Har bir `animal` uchun o'rtacha yoshni `df` da hisoblang.

In [ ]:
df.groupby('animal')['age'].mean()

animal
cat      2.333333
dog      5.000000
snake    2.500000
Name: age, dtype: float64

**16.** `df` yordamida har bir turdagi `animal` sonini hisoblang.

In [ ]:
df['animal'].value_counts()

animal
cat      4
dog      4
snake    2
Name: count, dtype: int64

**17.** `animal` ustunida, `snake` so'zini `python` so'ziga o'zgartiring.

In [ ]:
df['animal'] = df['animal'].replace('snake', 'python')
df

animal  age  visits priority
a     cat  2.5       1      yes
b     cat  3.0       3      yes
c  python  0.5       2       no
d     dog  NaN       3      yes
e     dog  5.0       2       no
f     cat  1.5       3       no
g  python  4.5       1       no
h     cat  NaN       1      yes
i     dog  7.0       2       no
j     dog  3.0       1       no

## `📌DataFrame: Murakkab`

### Bir oz qiyinroq: to'g'ri javob olish uchun ikki yoki undan ortiq usullarni birlashtirish kerak bo'lishi mumkin

Oldingi bo'lim DataFrame-ning asosiy, ammo muhim operatsiyalari bo'yicha vazifalarni jamlagan edi. Quyida ma'lumotlarni kesishingiz kerak bo'lgan mashqlar mavjud.

**18.** Sizda butun sonlardan iborat `A` ustunli DataFrame `df` mavjud. Misol uchun:
```python
df = pd.DataFrame({'A': [1, 2, 2, 3, 4, 5, 5, 5, 6, 7, 7]})
```

Yuqoridagi qator bilan bir xil butun sonni o'z ichiga olgan qator yarating.


💡Sizda quyidagi qiymatlarni o'z ichiga olgan ustun paydo bo'lishi kerak:

```python
1, 2, 3, 4, 5, 6, 7
```

In [ ]:
df = pd.DataFrame({'A': [1, 2, 2, 3, 4, 5, 5, 5, 6, 7, 7]})

df.loc[df['A'].shift() != df['A']]
df.drop_duplicates(subset='A')

A
0  1
1  2
3  3
4  4
5  5
8  6
9  7

**19.** Raqamli qiymatlardan iborat DataFrame berildi

```python
df = pd.DataFrame(np.random.random(size=(5, 3))) # float qiymatlarining 5x3 o'lchovi
```

Qatordagi har bir elementdan satrning o'rtacha qiymatini qanday ayirish mumkin?

In [ ]:
df = pd.DataFrame(np.random.random(size=(5, 3)))

In [ ]:
df.describe()

0         1         2
count  5.000000  5.000000  5.000000
mean   0.618252  0.283190  0.576464
std    0.298884  0.135128  0.222293
min    0.125055  0.175831  0.404953
25%    0.613726  0.186698  0.409872
50%    0.657005  0.230344  0.546669
75%    0.787596  0.320814  0.570775
max    0.907877  0.502265  0.950051

In [ ]:
df

0         1         2
0  0.787596  0.320814  0.404953
1  0.125055  0.230344  0.409872
2  0.657005  0.175831  0.546669
3  0.907877  0.502265  0.570775
4  0.613726  0.186698  0.950051

In [ ]:
df.sub(df.mean(axis=1), axis=0)

0         1         2
0  0.283142 -0.183641 -0.099501
1 -0.130035 -0.024746  0.154781
2  0.197170 -0.284004  0.086834
3  0.247571 -0.158041 -0.089530
4  0.030234 -0.396794  0.366560

**20.** Sizda 10 ta haqiqiy raqamlar ustunidan iborat DataFrame mavjud:

``` python
df = pd.DataFrame(np.random.random(size=(5, 10)), columns=list('abcdefghij'))
```
Qaysi sonlar ustunining yig'indisi eng kichik ekanini hisoblang. U ustunning yorlig'ini qaytaring.

In [ ]:
df = pd.DataFrame(np.random.random(size=(5, 10)), columns=list('abcdefghij'))

df.sum().idxmin()

'h'

In [ ]:
df.sum().idxmin()

'h'

**21.** Quyida butun sonlardan iborat DataFrame `df` mavjud. Har bir satrga  5 ta yozuv `NaN` qiymatlari kiritilgan.

DataFrame ning har bir qatori uchun uchinchi `NaN` qiymatini o'z ichiga olgan ustunni toping.

Ustun yorliqlarini qaytarishingiz kerak: `e, c, d, h, d`

In [ ]:
nan = np.nan
data = [[0.04,  nan,  nan, 0.25,  nan, 0.43, 0.71, 0.51,  nan,  nan],
        [ nan,  nan,  nan, 0.04, 0.76,  nan,  nan, 0.67, 0.76, 0.16],
        [ nan,  nan, 0.5 ,  nan, 0.31, 0.4 ,  nan,  nan, 0.24, 0.01],
        [0.49,  nan,  nan, 0.62, 0.73, 0.26, 0.85,  nan,  nan,  nan],
        [ nan,  nan, 0.41,  nan, 0.05,  nan, 0.61,  nan, 0.48, 0.68]]

columns = list('abcdefghij')

df = pd.DataFrame(data, columns=columns)

# Javobni bu yerga yozing!
(df.isnull().cumsum(axis=1) ==3).idxmax(axis=1)
# (df.isnull().cumsum(axis=1) == 3).idxmax(axis=1)

0    e
1    c
2    d
3    h
4    d
dtype: object

**22.** DataFrame `grps` guruhlari ustuniga va `vals` butun qiymatlari ustuniga ega:

```python
df = pd.DataFrame({'grps': list('aaabbcaabcccbbc'),
                   'vals': [12,345,3,1,45,14,4,52,54,23,235,21,57,3,87]})
```
Har bir `grps` uchun uchta eng katta qiymatning yig‘indisini toping.

💡Javobni quyidagicha yakunlashingiz kerak:
```
grps
a    409
b    156
c    345
```

     'a   a  a b  b  c a a  b  c  c   c  b  b  c'
      12,345,3,1,45,14,4,52,54,23,235,21,57,3,87

In [ ]:
df = pd.DataFrame({'grps': list('aaabbcaabcccbbc'),
                   'vals': [12,345,3,1,45,14,4,52,54,23,235,21,57,3,87]})
result = df.groupby('grps')['vals'].nlargest(3).sum()
# df.groupby('grps')['vals'].nlargest(3).sum(level=0)

In [ ]:
df = pd.DataFrame({'grps': list('aaabbcaabcccbbc'),
                   'vals': [12,345,3,1,45,14,4,52,54,23,235,21,57,3,87]})

# Javobni bu yerga yozing!
df.groupby('grps')['vals'].nlargest(3).sum()

910

**28.** Quyida tuzilgan DataFrame `df` ikkita butun son ustuniga ega 'A' va 'B'. "A" dagi qiymatlar 1 dan 100 gacha (shu jumladan).

“A” dagi har bir 10 ta ketma-ket butun sonlar guruhi uchun (ya’ni `(0, 10]`, `(10, 20]`, ...) 'B' ustunidagi mos qiymatlar yig'indisini hisoblang.

Javob quyidagicha bo'lishi kerak:

```
A
(0, 10]      635
(10, 20]     360
(20, 30]     315
(30, 40]     306
(40, 50]     750
(50, 60]     284
(60, 70]     424
(70, 80]     526
(80, 90]     835
(90, 100]    852
```

In [ ]:
df = pd.DataFrame(np.random.RandomState(8765).randint(1, 101, size=(100, 2)), columns = ["A", "B"])

# Javobni bu yerga yozing!
df['A_bins'] = pd.cut(df['A'],bins = range(0,110,10))

result = df.groupby('A_bins')['B'].sum()
result

A_bins
(0, 10]      635
(10, 20]     360
(20, 30]     315
(30, 40]     306
(40, 50]     750
(50, 60]     284
(60, 70]     424
(70, 80]     526
(80, 90]     835
(90, 100]    852
Name: B, dtype: int64

**29.** 'X' butun son ustuni bo'lgan DataFrame:
```python
df = pd.DataFrame({'X': [7, 2, 0, 3, 4, 2, 5, 0, 3, 4]})
```

Har bir qiymat uchun farqni oldingi nolga qaytaring.

Sizda natija quyidagicha bo'lishi kerak:

```
[1, 2, 0, 1, 2, 3, 4, 0, 1, 2]
```

Buni yangi `Y` ustuni deb nomlang.

In [ ]:

df = pd.DataFrame({'X': [7, 2, 0, 3, 4, 2, 5, 0, 3, 4]})

# df['Y'] = (df['X'] != 0).cumsum() - 1

# df['Y'].tolist()


y_values = []
counter = 0

for value in df['X']:
    if value == 0:
        counter = 0
    else:
        counter += 1
    y_values.append(counter)

df['Y'] = y_values

print(df['Y'].tolist())

[1, 2, 0, 1, 2, 3, 4, 0, 1, 2]



##`📌Ma'lumotlarni tozalash`

Bu har doim sodir bo'ladi: kimdir sizga noto'g'ri tuzilgan satrlar, ro'yxatlar va yetishmayotgan ma'lumotlarni o'z ichiga olgan ma'lumotlarni beradi. Tahlilni davom ettirish uchun uni qanday tartibga keltirasiz?

**30.** Quyidagi jumboqni tasavvur qilishi uchun ushbu misolni DataFrame sifatida oling:

```python
df = pd.DataFrame({'From_To': ['LoNDon_paris', 'MAdrid_miLAN', 'londON_StockhOlm',
                               'Budapest_PaRis', 'Brussels_londOn'],
              'FlightNumber': [10045, np.nan, 10065, np.nan, 10085],
              'RecentDelays': [[23, 47], [], [24, 43, 87], [13], [67, 32]],
                   'Airline': ['KLM(!)', '<Air France> (12)', '(British Airways. )',
                               '12. Air France', '"Swiss Air"']})
```
Tartiblaganingizdan so'ng u quyidagi ko'rinishga ega bo'lishi kerak:

```
            From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris       10045.0      [23, 47]               KLM(!)
1      MAdrid_miLAN           NaN            []    <Air France> (12)
2  londON_StockhOlm       10065.0  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis           NaN          [13]       12. Air France
4   Brussels_londOn       10085.0      [67, 32]          "Swiss Air"
```


In [ ]:
df = pd.DataFrame({'From_To': ['LoNDon_paris', 'MAdrid_miLAN', 'londON_StockhOlm',
                               'Budapest_PaRis', 'Brussels_londOn'],
              'FlightNumber': [10045, np.nan, 10065, np.nan, 10085],
              'RecentDelays': [[23, 47], [], [24, 43, 87], [13], [67, 32]],
                   'Airline': ['KLM(!)', '<Air France> (12)', '(British Airways. )',
                               '12. Air France', '"Swiss Air"']})


**31.** **FlightNumber** ustunidagi ba'zi qiymatlar yo'q. Ular `NaN` ko'rinishida yozilgan. Bu raqamlar har bir qatorda 10 ga ko'payishi kerak, shuning uchun 10055 va 10075 ni qo'yish kerak. Ushbu yetishmayotgan raqamlarni to'ldirish uchun `df` ni o'zgartiring va ustunni butun son ustuniga aylantiring.

In [ ]:
df["FlightNumber"]=df['FlightNumber'].interpolate(method='linear').astype(int)
df

From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris         10045      [23, 47]               KLM(!)
1      MAdrid_miLAN         10055            []    <Air France> (12)
2  londON_StockhOlm         10065  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis         10075          [13]       12. Air France
4   Brussels_londOn         10085      [67, 32]          "Swiss Air"

**32.** **From\_To** ustunini ikkita alohida ustun ko'rinishida yozing.                                          

To'g'ri qiymatlarga ega `temp` deb nomlangan yangi Temporary DataFrame berish uchun har bir satrni pastki chiziq chegarasiga ajrating.                                                                     Ushbu Temporary DataFrame ustunlarini `From` va `To` deb nomlang.

In [ ]:
temp = df['From_To'].str.split('_', expand=True)
temp.columns = ['From', 'To']
temp


From         To
0    LoNDon      paris
1    MAdrid      miLAN
2    londON  StockhOlm
3  Budapest      PaRis
4  Brussels     londOn

In [ ]:
df.join(temp)
del df['From_To']
df = df.join(temp)

In [ ]:
df

FlightNumber  RecentDelays              Airline      From         To
0         10045      [23, 47]               KLM(!)    LoNDon      paris
1         10055            []    <Air France> (12)    MAdrid      miLAN
2         10065  [24, 43, 87]  (British Airways. )    londON  StockhOlm
3         10075          [13]       12. Air France  Budapest      PaRis
4         10085      [67, 32]          "Swiss Air"  Brussels     londOn


**34.** Temporary DataFrame `temp` da shahar nomlari betartib yozilgan.Satrlarning faqat birinchi harfi katta yozilishi kerak.
                                                                         (Masalan, "londON" "London" bo'lishi kerak.)

In [ ]:
# temp['From'].map(lambda x: x.title())
temp[['From','To']] = temp[['From','To']].applymap(lambda x: x.title())
temp

From         To
0    London      Paris
1    Madrid      Milan
2    London  Stockholm
3  Budapest      Paris
4  Brussels     London

**33**. `Airline` ustunida aviakompaniya nomlari atrofida qoʻshimcha tinish belgilari va belgilar paydo boʻlganini koʻrishingiz mumkin. Faqat aviakompaniya nomini qoldiring. Masalan, ʻ'(British Airways. )'' `British Airways` ko'rinishida yozilishi kerak.

In [ ]:
import re
df['Airline'] = df['Airline'].apply(lambda x: re.sub(r'[^\w\s]','',x))
df

FlightNumber  RecentDelays           Airline      From         To
0         10045      [23, 47]               KLM    LoNDon      paris
1         10055            []     Air France 12    MAdrid      miLAN
2         10065  [24, 43, 87]  British Airways     londON  StockhOlm
3         10075          [13]     12 Air France  Budapest      PaRis
4         10085      [67, 32]         Swiss Air  Brussels     londOn

**34.** `df` dan **From_To** ustunini o'chiring va Temporary DataFrame `temp` ustunini qo'shing.

In [ ]:
# df = df.join(temp)
df[['From','To']] = df[['From','To']].applymap(lambda x: x.title())
df

FlightNumber  RecentDelays           Airline      From         To
0         10045      [23, 47]               KLM    London      Paris
1         10055            []     Air France 12    Madrid      Milan
2         10065  [24, 43, 87]  British Airways     London  Stockholm
3         10075          [13]     12 Air France  Budapest      Paris
4         10085      [67, 32]         Swiss Air  Brussels     London

**34**. `RecentDelays` ustunida qiymatlar DataFrame-ga ro'yxat sifatida kiritilgan. Sizning vazifangiz birinchi qiymatni o'z ustunida, har bir ikkinchi qiymatni o'z ustunida va qolgan qiymatlar ham ustunga mos holda chop etish. Agar N-chi qiymat bo'lmasa, qiymat `NaN` deb nomlanishi kerak.

Roʻyxatlar qatorini `delays` nomli DataFrame ichiga kengaytiring, `delay_1`, `delay_2` va qolgan ustunlarni ham nomlarini oʻzgartiring va `df` dagi keraksiz `RecentDelays` ustunini `delays` bilan almashtiring.

In [ ]:
# df['RecentDelays']
delays = pd.DataFrame(
    df['RecentDelays'].values.tolist(),
    columns=['delay_{}'.format(i+1) for i in range(df['RecentDelays'].str.len().max())]
)
delays

delay_1  delay_2  delay_3
0     23.0     47.0      NaN
1      NaN      NaN      NaN
2     24.0     43.0     87.0
3     13.0      NaN      NaN
4     67.0     32.0      NaN

In [ ]:
df = df.drop('RecentDelays', axis =1)
df = pd.concat([df,delays], axis =1)
df

FlightNumber           Airline      From         To  delay_1  delay_2  \
0         10045               KLM    London      Paris     23.0     47.0   
1         10055     Air France 12    Madrid      Milan      NaN      NaN   
2         10065  British Airways     London  Stockholm     24.0     43.0   
3         10075     12 Air France  Budapest      Paris     13.0      NaN   
4         10085         Swiss Air  Brussels     London     67.0     32.0   

   delay_3  
0      NaN  
1      NaN  
2     87.0  
3      NaN  
4      NaN

Natija quyidagi ko'rinishda bo'lishi kerak:
```
   FlightNumber          Airline      From         To  delay_1  delay_2  delay_3
0         10045              KLM    London      Paris     23.0     47.0      NaN
1         10055       Air France    Madrid      Milan      NaN      NaN      NaN
2         10065  British Airways    London  Stockholm     24.0     43.0     87.0
3         10075       Air France  Budapest      Paris     13.0      NaN      NaN
4         10085        Swiss Air  Brussels     London     67.0     32.0      NaN
```

In [ ]:
df

FlightNumber           Airline      From         To  delay_1  delay_2  \
0         10045               KLM    London      Paris     23.0     47.0   
1         10055     Air France 12    Madrid      Milan      NaN      NaN   
2         10065  British Airways     London  Stockholm     24.0     43.0   
3         10075     12 Air France  Budapest      Paris     13.0      NaN   
4         10085         Swiss Air  Brussels     London     67.0     32.0   

   delay_3  
0      NaN  
1      NaN  
2     87.0  
3      NaN  
4      NaN

##`📌Minesweeper`


### Minesweeper tarmog'ida xavfsiz kvadratlar uchun raqamlar yaratamiz!

💡 Agar siz Windowsning avvalgi versiyalaridan foydalangan bo'lsangiz, Minesweeper bilan tanishsiz:
- https://en.wikipedia.org/wiki/Minesweeper_(video_game)

Agar siz o'yin bilan tanish bo'lmasangiz, kvadratchalar to'plamini tasavvur qiling: bu kvadratlarning ba'zilari mineni yashiradi. Agar siz mineni bossangiz, darhol yo'qotasiz. Agar siz xavfsiz kvadratni bossangiz, darhol qo'shni bo'lgan kvadratchalarda qancha mine borligini ko'rsatadigan raqamni ko'rasiz. O'yinning maqsadi to'rdagi mine bo'lmagan barcha kvadratlarni ochishdir.

Ushbu bo'limda biz Minesweeper o'yini uchun kerakli ma'lumotlarni o'z ichiga olgan DataFrame yaratamiz: kvadratlarning koordinatalari, kvadratda mine bormi va qo'shni kvadratlarda topilgan minelar sonini aniqlaymiz!

**35**. Tasavvur qiling, biz Minesweeper o'yinini 5 dan 4 gacha bo'lgan panjarada o'ynayapmiz, ya'ni.
```
X = 5
Y = 4
```
Boshlash uchun ikkita ustunli DataFrame `df` yarating, bu tarmoq uchun har bir koordinatani o'z ichiga olgan `'x'` va `'y'`. Ya'ni DataFrame ishga tushishi kerak:
```
    x y
0 0 0
1 0 1
2 0 2
```

In [ ]:
X =5
Y =4
coords = [(x,y) for y in range(Y) for x in range(X)]
df = pd.DataFrame(coords, columns = ['x','y'])
df

x  y
0   0  0
1   1  0
2   2  0
3   3  0
4   4  0
5   0  1
6   1  1
7   2  1
8   3  1
9   4  1
10  0  2
11  1  2
12  2  2
13  3  2
14  4  2
15  0  3
16  1  3
17  2  3
18  3  3
19  4  3

**36**. Ushbu DataFrame `df` uchun yangi nol (safe) va birliklar (mine) ustunini yarating. Har bir joyda mine sodir bo'lish ehtimoli 0,4 bo'lishi kerak.

In [ ]:
df['value'] = np.where(np.random.rand(len(df))<0.4,1,0)
df

x  y  value
0   0  0      1
1   1  0      0
2   2  0      0
3   3  0      1
4   4  0      1
5   0  1      1
6   1  1      1
7   2  1      0
8   3  1      1
9   4  1      0
10  0  2      0
11  1  2      0
12  2  2      0
13  3  2      0
14  4  2      1
15  0  3      0
16  1  3      0
17  2  3      1
18  3  3      0
19  4  3      0

**37**. Endi ushbu DataFrame uchun `adjacent` nomli yangi ustun yarating. Ushbu ustunda tarmoqdagi qo'shni kvadratlarda topilgan minelar soni bo'lishi kerak.

(Masalan, `(0, 0)` koordinatasi uchun yozuv bo'lgan birinchi qator uchun `(0, 1)`, `(1, 0)` va `(1) koordinatalarida qancha mine borligini hisoblang. , 1)`.)

In [ ]:
adjacent_column = []

for i in range(len(df) - 1):
    adjacent_mines = sum([1 for j in range(len(df.columns) - 1) if not pd.isna(df.iloc[i, j])])
    adjacent_column.append(adjacent_mines)

# O'xshash bo'lgan eng oxirgi qiymatni qo'shib qo'yish
adjacent_column.append(adjacent_column[-1])

# Yangi ustun qo'shish
df['Adjacent'] = adjacent_column


x  y  value  Adjacent
0   0  0      1         3
1   1  0      0         3
2   2  0      0         3
3   3  0      1         3
4   4  0      1         3
5   0  1      1         3
6   1  1      1         3
7   2  1      0         3
8   3  1      1         3
9   4  1      0         3
10  0  2      0         3
11  1  2      0         3
12  2  2      0         3
13  3  2      0         3
14  4  2      1         3
15  0  3      0         3
16  1  3      0         3
17  2  3      1         3
18  3  3      0         3
19  4  3      0         3

**38**. Mineni o'z ichiga olgan DataFrame qatorlari uchun `adjacent` ustunidagi qiymatni `NaN` ga o'zgartiring.

In [ ]:
adjacent_column = []

for i in range(len(df) - 1):
    adjacent_mines = sum([1 for j in range(len(df.columns) - 1) if not pd.isna(df.iloc[i, j])])
    adjacent_column.append(adjacent_mines)

# O'xshash bo'lgan eng oxirgi qiymatni qo'shib qo'yish
adjacent_column.append(adjacent_column[-1])

# Yangi ustun qo'shish
df['Adjacent'] = adjacent_column

# Adjacent ustunidagi qiymatlarni NaN ga o'zgartirish
df.loc[:, 'Adjacent'] = np.nan

df

x  y  value  Adjacent
0   0  0      1       NaN
1   1  0      0       NaN
2   2  0      0       NaN
3   3  0      1       NaN
4   4  0      1       NaN
5   0  1      1       NaN
6   1  1      1       NaN
7   2  1      0       NaN
8   3  1      1       NaN
9   4  1      0       NaN
10  0  2      0       NaN
11  1  2      0       NaN
12  2  2      0       NaN
13  3  2      0       NaN
14  4  2      1       NaN
15  0  3      0       NaN
16  1  3      0       NaN
17  2  3      1       NaN
18  3  3      0       NaN
19  4  3      0       NaN

**39**. DataFrame-ni `adjacent` minelar sonining panjarasiga aylantiring: ustunlar `x` koordinatasi, qatorlar `y` koordinatasi.

In [ ]:
df['Adjacent'] = adjacent_column

# Adjacent ustunidagi qiymatlarni NaN ga o'zgartirish
df.loc[:, 'Adjacent'] = np.nan

# DataFrame-ni aylantirish
df.index = range(len(df), 0, -1)  # Qatorlar indeksini alohida o'rnating
df.columns = range(1, len(df.columns) + 1)  # Ustunlar indeksini alohida o'rnating


##📌BONUS!

O'zingiz javob berishni istagan savolni kiriting va unga javob bering!